File: pulsar-client-cpp/tests/MemoryLimitTest.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <gtest/gtest.h>
#include <array>
#include <thread>

#include "../lib/MemoryLimitController.h"
#include "../lib/Latch.h"
#include "../lib/Future.h"
#include "../lib/Utils.h"

#include <pulsar/Client.h>

using namespace pulsar;

extern std::string lookupUrl;
extern std::string unique_str();

TEST(MemoryLimitTest, testRejectMessages) {
    std::string topic = "topic-" + unique_str();

    ClientConfiguration config;
    config.setMemoryLimit(100 * 1024);
    Client client(lookupUrl, config);

    ProducerConfiguration producerConf;
    producerConf.setBlockIfQueueFull(false);
    Producer producer;
    Result res = client.createProducer(topic, producerConf, producer);
    ASSERT_EQ(res, ResultOk);

    const int n = 101;
    Latch latch(n);

    std::array<char, 1024> buffer;

    for (int i = 0; i < n; i++) {
        producer.sendAsync(MessageBuilder().setContent(buffer.data(), buffer.size()).build(),
                           [&](Result res, const MessageId& msgId) {
                               ASSERT_EQ(res, ResultOk);
                               latch.countdown();
                           });
    }

    res = producer.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultMemoryBufferIsFull);

    latch.wait();

    // We should now be able to send again
    res = producer.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultOk);
}

TEST(MemoryLimitTest, testRejectMessagesOnMultipleTopics) {
    std::string t1 = "topic-1-" + unique_str();
    std::string t2 = "topic-2-" + unique_str();

    ClientConfiguration config;
    config.setMemoryLimit(100 * 1024);
    Client client(lookupUrl, config);

    ProducerConfiguration producerConf;
    producerConf.setBlockIfQueueFull(false);
    producerConf.setBatchingMaxPublishDelayMs(10000);

    Producer p1;
    Result res = client.createProducer(t1, producerConf, p1);
    ASSERT_EQ(res, ResultOk);

    Producer p2;
    res = client.createProducer(t2, producerConf, p2);
    ASSERT_EQ(res, ResultOk);

    const int n = 101;
    Latch latch(n);

    std::array<char, 1024> buffer;

    for (int i = 0; i < n / 2; i++) {
        p1.sendAsync(MessageBuilder().setContent(buffer.data(), buffer.size()).build(),
                     [&](Result res, const MessageId& msgId) {
                         ASSERT_EQ(res, ResultOk);
                         latch.countdown();
                     });

        p2.sendAsync(MessageBuilder().setContent(buffer.data(), buffer.size()).build(),
                     [&](Result res, const MessageId& msgId) {
                         ASSERT_EQ(res, ResultOk);
                         latch.countdown();
                     });
    }

    // Last message in order to reach the limit
    p1.sendAsync(MessageBuilder().setContent(buffer.data(), buffer.size()).build(),
                 [&](Result res, const MessageId& msgId) {
                     ASSERT_EQ(res, ResultOk);
                     latch.countdown();
                 });

    res = p1.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultMemoryBufferIsFull);

    res = p2.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultMemoryBufferIsFull);

    latch.wait();

    // We should now be able to send again
    res = p1.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultOk);

    res = p2.send(MessageBuilder().setContent(buffer.data(), buffer.size()).build());
    ASSERT_EQ(res, ResultOk);
}

TEST(MemoryLimitTest, testNoProducerQueueSize) {
    std::string topic = "topic-" + unique_str();

    ClientConfiguration config;
    config.setMemoryLimit(10 * 1024);
    Client client(lookupUrl, config);

    ProducerConfiguration producerConf;
    producerConf.setBlockIfQueueFull(true);
    producerConf.setMaxPendingMessages(0);
    producerConf.setMaxPendingMessagesAcrossPartitions(0);
    Producer producer;
    Result res = client.createProducer(topic, producerConf, producer);
    ASSERT_EQ(res, ResultOk);

    std::array<Promise<Result, MessageId>, 100> promises;

    for (int i = 0; i < 100; i++) {
        producer.sendAsync(MessageBuilder().setContent("hello").build(),
                           WaitForCallbackValue<MessageId>(promises[i]));
    }

    producer.flush();

    for (auto& p : promises) {
        MessageId id;
        Result res = p.getFuture().get(id);
        ASSERT_EQ(res, ResultOk);
    }
}

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/MemoryLimitTest.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.pulsar.client.api;

import lombok.Cleanup;
import org.apache.pulsar.client.api.PulsarClientException.MemoryBufferIsFullError;
import org.apache.pulsar.client.impl.ProducerImpl;
import org.apache.pulsar.client.impl.PulsarTestClient;
import org.awaitility.Awaitility;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.time.Duration;
import java.util.concurrent.TimeUnit;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.fail;

@Test(groups = "broker-api")
public class MemoryLimitTest extends ProducerConsumerBase {

    @DataProvider(name = "batching")
    public Object[][] provider() {
        return new Object[][] {
                // "Batching"
                { false },
                { true },
        };
    }

    @BeforeMethod
    @Override
    protected void setup() throws Exception {
        super.internalSetup();
        super.producerBaseSetup();
    }

    @AfterMethod(alwaysRun = true)
    @Override
    protected void cleanup() throws Exception {
        super.internalCleanup();
    }

    @Test
    public void testRejectMessages()
            throws Exception {
        String topic = newTopicName();

        ClientBuilder clientBuilder = PulsarClient.builder()
                .serviceUrl(pulsar.getBrokerServiceUrl())
                .memoryLimit(100, SizeUnit.KILO_BYTES);

        @Cleanup
        PulsarTestClient client = PulsarTestClient.create(clientBuilder);

        @Cleanup
        ProducerImpl<byte[]> producer = (ProducerImpl<byte[]>) client.newProducer()
                .topic(topic)
                .blockIfQueueFull(false)
                .sendTimeout(5, TimeUnit.SECONDS)
                .create();

        // make sure all message pending at pendingMessages queue
        // connection with broker can not be established, so handleSendReceipt will not be invoked while sending message
        client.dropOpSendMessages();
        final int n = 101;
        for (int i = 0; i < n; i++) {
            producer.sendAsync(new byte[1024]);
        }
        Awaitility.await()
                .atMost(Duration.ofSeconds(5))
                .until(() -> producer.getPendingQueueSize() == n);
        assertEquals(client.getMemoryLimitController().currentUsage(), n * 1024);

        try {
            producer.send(new byte[1024]);
            fail("should have failed");
        } catch (MemoryBufferIsFullError e) {
            // Expected
        }

        client.allowReconnecting();
        Awaitility.await()
                .atMost(Duration.ofSeconds(30))
                .until(() -> producer.getPendingQueueSize() == 0);
        assertEquals(client.getMemoryLimitController().currentUsage(), 0);

        // We should now be able to send again
        producer.send(new byte[1024]);
    }

    @Test
    public void testRejectMessagesOnMultipleTopics() throws Exception {
        String t1 = newTopicName();
        String t2 = newTopicName();

        ClientBuilder clientBuilder = PulsarClient.builder()
                .serviceUrl(pulsar.getBrokerServiceUrl())
                .memoryLimit(100, SizeUnit.KILO_BYTES);

        @Cleanup
        PulsarTestClient client = PulsarTestClient.create(clientBuilder);

        @Cleanup
        ProducerImpl<byte[]> p1 = (ProducerImpl<byte[]>) client.newProducer()
                .topic(t1)
                .blockIfQueueFull(false)
                .sendTimeout(5, TimeUnit.SECONDS)
                .create();

        @Cleanup
        ProducerImpl<byte[]> p2 = (ProducerImpl<byte[]>) client.newProducer()
                .topic(t2)
                .blockIfQueueFull(false)
                .sendTimeout(5, TimeUnit.SECONDS)
                .create();

        client.dropOpSendMessages();
        final int n = 101;
        for (int i = 0; i < n / 2; i++) {
            p1.sendAsync(new byte[1024]);
            p2.sendAsync(new byte[1024]);
        }

        // Last message in order to reach the limit
        p1.sendAsync(new byte[1024]);

        Awaitility.await()
                .atMost(Duration.ofSeconds(5))
                .until(() -> (p1.getPendingQueueSize() + p2.getPendingQueueSize()) == n);
        assertEquals(client.getMemoryLimitController().currentUsage(), n * 1024);

        try {
            p1.send(new byte[1024]);
            fail("should have failed");
        } catch (MemoryBufferIsFullError e) {
            // Expected
        }

        try {
            p2.send(new byte[1024]);
            fail("should have failed");
        } catch (MemoryBufferIsFullError e) {
            // Expected
        }

        client.allowReconnecting();
        Awaitility.await()
                .atMost(Duration.ofSeconds(30))
                .until(() -> (p1.getPendingQueueSize() + p2.getPendingQueueSize()) == 0);
        assertEquals(client.getMemoryLimitController().currentUsage(), 0);

        // We should now be able to send again
        p1.send(new byte[1024]);
        p2.send(new byte[1024]);
    }
}
